src/main/java/bgu/spl/mics/application/subscribers/M.java

```diff
@@ -1,24 +1,103 @@
- * You MAY change constructor signatures and even add new public constructors.
+ * <p>
+ * You can add private fields and public methods to this class. You MAY change constructor signatures and even add new
+ * public constructors.
  */
 public class M extends Subscriber {
+    int _currTime = 0;
+    int _serial;
+    Diary diary;
+
+    public M(int serial) {
+        super("M" + serial);
+        _serial = serial;
+        diary = Diary.getInstance();
+    }
 
-	public M() {
-		super("Change_This_Name");
-		// TODO Implement this
-	}
+    @Override
+    protected void initialize() {
+        subscribeBroadcast(TickBroadcast.class, (TickBroadcast b) -> _currTime = b.getCurrTime());
+        subscribeEvent(MissionReceivedEvent.class, (MissionReceivedEvent e) -> {
+            diary.incrementTotal();
+            // get all mission details
+            MissionInfo info = e.getMission();
+            String missionName = info.getMissionName();
+            int timeIssued = info.getTimeIssued();
+            List<String> agentsSerials = info.getSerialAgentsNumbers();
+            // ask for agents from Moneypenny
+            AgentsAvailableObject agentsAvailableObject = new AgentsAvailableObject(agentsSerials, info.getDuration());
+            Future<AgentsAvailableObject> agentsAvailableFuture;
+            agentsAvailableFuture = this.getSimplePublisher().sendEvent(new AgentsAvailableEvent(agentsAvailableObject));
+            // check whether Moneypenny acquired the agents, and allocate all relevant details
+            agentsAvailableObject = agentsAvailableFuture.get(); // M and Moneypenny uses the same agentsObject
+            List<String> agentNames = agentsAvailableObject.getAgentsNames();
+            if (agentNames.get(0).equals("")) return;
+            int MoneypennySerial = agentsAvailableObject.getMoneypennySerial();
+            // ask for gadget from Q
+            GadgetAvailableObject gadgetAvailableObject = new GadgetAvailableObject(info.getGadget());
+            Future<GadgetAvailableObject> gadgetAvailableFuture;
+            gadgetAvailableFuture = this.getSimplePublisher().sendEvent(new GadgetAvailableEvent(gadgetAvailableObject));
+            // check whether Q acquired the gadget, and allocate all relevant details
+            gadgetAvailableObject = gadgetAvailableFuture.get(); //M and Q uses the same gadgetObject
+            if (!gadgetAvailableObject.isGadgetExists()) return;
+            String gadget = gadgetAvailableObject.getGadget();
+            int Qtime = gadgetAvailableObject.getQtime();
+            // check if mission time expired
+            if (_currTime > info.getTimeExpired()) {
+                // signal Moneypenny she should release the agents.
+                agentsAvailableObject.terminateMission();
+                return;
+            }
+            // signal Moneypenny she should send the agents.
+            agentsAvailableObject.sendMission();
 
```

Comment:
Shouldn't there be some sort of signaling to Moneypenny that something changed? maybe a `agentsAvailableObject.notifyAll()` or something